### PR TITLE
Add compiler warning for inferred tailcalls in TCO'd cycles

### PR DESCRIPTION
**Pull request by tylerhou**
The `-analyze-tailcall-cfg` flag now emits warnings for {function applications in tail position that are inferred to be nontail} that also may participate in a cycle of tailcalls. Such cycles (tail edges with one inferred nontail edge) are dangerous because traversing the cycle repeatedly may cause the program to stack overflow. Example:

```ocaml
let rec collatz n k =
  let[@inline never] next n k =
    let n = n - 1
    and k = if k mod 2 == 0 then (k / 2) else ((3 * k) + 1) in
    collatz n k
  in
  ignore next;
  if n > 0 then next n k else k


File "warnings.ml", line 29, characters 16-24:
29 |   if n > 0 then next n k else k
                     ^^^^^^^^
Warning 186 [inferred-nontail-in-tcod-cycle]: This function
call was inferred as nontail, but it may participate in a
cycle of TCO'd tail calls, which could overflow the stack at
runtime. Consider marking the function call with [@tail].
```

Currently, because this analysis is (mostly) conservative, and our infer tail/nontail heuristic is not precise, many such warnings are spurious.

We have also added some more test cases in `dcfg_tailcalls.ml` and changed the visualization so it renders compiler-synthesized edges with dashes. Here is the new graph:

image

### CODE REVIEW

File: backend/asmgen.ml [hunk 1]
```diff
@@ -274,6 +274,15 @@ let compile_fundecl ~ppf_dump ~funcnames fd_cmm =
...
   ++ Profile.record ~accumulate:true "cmm_invariants" (cmm_invariants ppf_dump)
+  ++ (fun fd ->
+      match !Flambda_backend_flags.cfg_analyze_tailcalls
+            || !Flambda_backend_flags.dcfg_tailcalls
+      with
+      | false -> fd
+      | true ->
+          fd ++ Analyze_tailcall_graph.fixup_inlined_tailcalls)
+          (* ++ Profile.record ~accumulate:true "fixup_inlined_tailcalls" *)
+          (*      Analyze_tailcall_graph.fixup_inlined_tailcalls) *)
```
Comment (dvulakh) [inline backend/asmgen.ml:285]: Let's return this.

**Review by dvulakh — CHANGES_REQUESTED** [review]
Looks good overall! Left a few comments.

File: backend/analyze_tailcall_graph.ml [hunk 2]
```diff
@@ -197,51 +210,86 @@ end = struct
...
+         "duplicate" edges. E.g. if A calls B in tail position multiple times in
+         different locations, we will record one edge for each call. (Each call
+         should be treated as distinct when emitting warnings.
+
+         Edges from Unknown_fn are not stored in here, since we don't want/need
+         to emit warnings for them, and they don't have a location.
+
+         Vertex.Tbl is a multimap, so this is really a map from a vertex to a
+         list of its edges. *)
+      adjacencies_with_loc : Edge.with_loc Vertex.Tbl.t
```
[the remark this answers is not in the document]
Comment (tylerhou) [reply backend/analyze_tailcall_graph.ml:247]: Done.

File: backend/analyze_tailcall_graph.ml [hunk 3]
```diff
@@ -507,4 +593,71 @@ module Global_state = struct
...
+    let nontail = fix ~tail_pos_ctx:false in
+    match expr with
+    | Cop (op, exprs, dbg) ->
+      let op : Cmm.operation =
+        (* Warning 4 is [fragile-match] *)
+        match[@ocaml.warning "-4"] op with
+        | Capply (machtype, region_close, original_position) ->
+          let inlined_position : Lambda.position_and_tail_attribute =
+            match original_position, tail_pos_ctx with
+            | Tail_position _, false ->
```
[the remark this answers is not in the document]
Comment (tylerhou) [reply backend/analyze_tailcall_graph.ml:614]: Done.

File: backend/asmgen.ml (hunk 1 above)
Comment (tylerhou) [reply backend/asmgen.ml:285]: Done.

**Review by dvulakh — APPROVED** [review]
Looks good, thanks!

**Pull request merged**